// @open-wallet-standard/core/index.js
/* tslint:disable */
/* eslint-disable */
/* prettier-ignore */

/* auto-generated by NAPI-RS */

const { existsSync, readFileSync } = require('fs')
const { join } = require('path')

const { platform, arch } = process

let nativeBinding = null
let localFileExisted = false
let loadError = null

function isMusl() {
  // For Node 10
  if (!process.report || typeof process.report.getReport !== 'function') {
    try {
      const lddPath = require('child_process').execSync('which ldd').toString().trim()
      return readFileSync(lddPath, 'utf8').includes('musl')
    } catch (e) {
      return true
    }
  } else {
    const { glibcVersionRuntime } = process.report.getReport().header
    return !glibcVersionRuntime
  }
}

switch (platform) {
  case 'android':
    switch (arch) {
      case 'arm64':
        localFileExisted = existsSync(join(__dirname, 'ows-node.android-arm64.node'))
        try {
          if (localFileExisted) {
            nativeBinding = require('./ows-node.android-arm64.node')
          } else {
            nativeBinding = require('@open-wallet-standard/core-android-arm64')
          }
        } catch (e) {
          loadError = e
        }
        break
      case 'arm':
        localFileExisted = existsSync(join(__dirname, 'ows-node.android-arm-eabi.node'))
        try {
          if (localFileExisted) {
            nativeBinding = require('./ows-node.android-arm-eabi.node')
          } else {
            nativeBinding = require('@open-wallet-standard/core-android-arm-eabi')
          }
        } catch (e) {
          loadError = e
        }
        break
      default:
        throw new Error(`Unsupported architecture on Android ${arch}`)
    }
    break
  case 'win32':
    switch (arch) {
      case 'x64':
        localFileExisted = existsSync(
          join(__dirname, 'ows-node.win32-x64-msvc.node')
        )
        try {
          if (localFileExisted) {
            nativeBinding = require('./ows-node.win32-x64-msvc.node')
          } else {
            nativeBinding = require('@open-wallet-standard/core-win32-x64-msvc')
          }
        } catch (e) {
          loadError = e
        }
        break
      case 'ia32':
        localFileExisted = existsSync(
          join(__dirname, 'ows-node.win32-ia32-msvc.node')
        )
        try {
          if (localFileExisted) {
            nativeBinding = require('./ows-node.win32-ia32-msvc.node')
          } else {
            nativeBinding = require('@open-wallet-standard/core-win32-ia32-msvc')
          }
        } catch (e) {
          loadError = e
        }
        break
      case 'arm64':
        localFileExisted = existsSync(
          join(__dirname, 'ows-node.win32-arm64-msvc.node')
        )
        try {
          if (localFileExisted) {
            nativeBinding = require('./ows-node.win32-arm64-msvc.node')
          } else {
            nativeBinding = require('@open-wallet-standard/core-win32-arm64-msvc')
          }
        } catch (e) {
          loadError = e
        }
        break
      default:
        throw new Error(`Unsupported architecture on Windows: ${arch}`)
    }
    break
  case 'darwin':
    localFileExisted = existsSync(join(__dirname, 'ows-node.darwin-universal.node'))
    try {
      if (localFileExisted) {
        nativeBinding = require('./ows-node.darwin-universal.node')
      } else {
        nativeBinding = require('@open-wallet-standard/core-darwin-universal')
      }
      break
    } catch {}
    switch (arch) {
      case 'x64':
        localFileExisted = existsSync(join(__dirname, 'ows-node.darwin-x64.node'))
        try {
          if (localFileExisted) {
            nativeBinding = require('./ows-node.darwin-x64.node')
          } else {
            nativeBinding = require('@open-wallet-standard/core-darwin-x64')
          }
        } catch (e) {
          loadError = e
        }
        break
      case 'arm64':
        localFileExisted = existsSync(
          join(__dirname, 'ows-node.darwin-arm64.node')
        )
        try {
          if (localFileExisted) {
            nativeBinding = require('./ows-node.darwin-arm64.node')
          } else {
            nativeBinding = require('@open-wallet-standard/core-darwin-arm64')
          }
        } catch (e) {
          loadError = e
        }
        break
      default:
        throw new Error(`Unsupported architecture on macOS: ${arch}`)
    }
    break
  case 'freebsd':
    if (arch !== 'x64') {
      throw new Error(`Unsupported architecture on FreeBSD: ${arch}`)
    }
    localFileExisted = existsSync(join(__dirname, 'ows-node.freebsd-x64.node'))
    try {
      if (localFileExisted) {
        nativeBinding = require('./ows-node.freebsd-x64.node')
      } else {
        nativeBinding = require('@open-wallet-standard/core-freebsd-x64')
      }
    } catch (e) {
      loadError = e
    }
    break
  case 'linux':
    switch (arch) {
      case 'x64':
        if (isMusl()) {
          localFileExisted = existsSync(
            join(__dirname, 'ows-node.linux-x64-musl.node')
          )
          try {
            if (localFileExisted) {
              nativeBinding = require('./ows-node.linux-x64-musl.node')
            } else {
              nativeBinding = require('@open-wallet-standard/core-linux-x64-musl')
            }
          } catch (e) {
            loadError = e
          }
        } else {
          localFileExisted = existsSync(
            join(__dirname, 'ows-node.linux-x64-gnu.node')
          )
          try {
            if (localFileExisted) {
              nativeBinding = require('./ows-node.linux-x64-gnu.node')
            } else {
              nativeBinding = require('@open-wallet-standard/core-linux-x64-gnu')
            }
          } catch (e) {
            loadError = e
          }
        }
        break
      case 'arm64':
        if (isMusl()) {
          localFileExisted = existsSync(
            join(__dirname, 'ows-node.linux-arm64-musl.node')
          )
          try {
            if (localFileExisted) {
              nativeBinding = require('./ows-node.linux-arm64-musl.node')
            } else {
              nativeBinding = require('@open-wallet-standard/core-linux-arm64-musl')
            }
          } catch (e) {
            loadError = e
          }
        } else {
          localFileExisted = existsSync(
            join(__dirname, 'ows-node.linux-arm64-gnu.node')
          )
          try {
            if (localFileExisted) {
              nativeBinding = require('./ows-node.linux-arm64-gnu.node')
            } else {
              nativeBinding = require('@open-wallet-standard/core-linux-arm64-gnu')
            }
          } catch (e) {
            loadError = e
          }
        }
        break
      case 'arm':
        if (isMusl()) {
          localFileExisted = existsSync(
            join(__dirname, 'ows-node.linux-arm-musleabihf.node')
          )
          try {
            if (localFileExisted) {
              nativeBinding = require('./ows-node.linux-arm-musleabihf.node')
            } else {
              nativeBinding = require('@open-wallet-standard/core-linux-arm-musleabihf')
            }
          } catch (e) {
            loadError = e
          }
        } else {
          localFileExisted = existsSync(
            join(__dirname, 'ows-node.linux-arm-gnueabihf.node')
          )
          try {
            if (localFileExisted) {
              nativeBinding = require('./ows-node.linux-arm-gnueabihf.node')
            } else {
              nativeBinding = require('@open-wallet-standard/core-linux-arm-gnueabihf')
            }
          } catch (e) {
            loadError = e
          }
        }
        break
      case 'riscv64':
        if (isMusl()) {
          localFileExisted = existsSync(
            join(__dirname, 'ows-node.linux-riscv64-musl.node')
          )
          try {
            if (localFileExisted) {
              nativeBinding = require('./ows-node.linux-riscv64-musl.node')
            } else {
              nativeBinding = require('@open-wallet-standard/core-linux-riscv64-musl')
            }
          } catch (e) {
            loadError = e
          }
        } else {
          localFileExisted = existsSync(
            join(__dirname, 'ows-node.linux-riscv64-gnu.node')
          )
          try {
            if (localFileExisted) {
              nativeBinding = require('./ows-node.linux-riscv64-gnu.node')
            } else {
              nativeBinding = require('@open-wallet-standard/core-linux-riscv64-gnu')
            }
          } catch (e) {
            loadError = e
          }
        }
        break
      case 's390x':
        localFileExisted = existsSync(
          join(__dirname, 'ows-node.linux-s390x-gnu.node')
        )
        try {
          if (localFileExisted) {
            nativeBinding = require('./ows-node.linux-s390x-gnu.node')
          } else {
            nativeBinding = require('@open-wallet-standard/core-linux-s390x-gnu')
          }
        } catch (e) {
          loadError = e
        }
        break
      default:
        throw new Error(`Unsupported architecture on Linux: ${arch}`)
    }
    break
  default:
    throw new Error(`Unsupported OS: ${platform}, architecture: ${arch}`)
}

if (!nativeBinding) {
  if (loadError) {
    throw loadError
  }
  throw new Error(`Failed to load native binding`)
}

const { generateMnemonic, deriveAddress, createWallet, importWalletMnemonic, importWalletPrivateKey, listWallets, getWallet, deleteWallet, exportWallet, renameWallet, signTransaction, signMessage, signTypedData, createPolicy, listPolicies, getPolicy, deletePolicy, createApiKey, listApiKeys, revokeApiKey, signAndSend } = nativeBinding

module.exports.generateMnemonic = generateMnemonic
module.exports.deriveAddress = deriveAddress
module.exports.createWallet = createWallet
module.exports.importWalletMnemonic = importWalletMnemonic
module.exports.importWalletPrivateKey = importWalletPrivateKey
module.exports.listWallets = listWallets
module.exports.getWallet = getWallet
module.exports.deleteWallet = deleteWallet
module.exports.exportWallet = exportWallet
module.exports.renameWallet = renameWallet
module.exports.signTransaction = signTransaction
module.exports.signMessage = signMessage
module.exports.signTypedData = signTypedData
module.exports.createPolicy = createPolicy
module.exports.listPolicies = listPolicies
module.exports.getPolicy = getPolicy
module.exports.deletePolicy = deletePolicy
module.exports.createApiKey = createApiKey
module.exports.listApiKeys = listApiKeys
module.exports.revokeApiKey = revokeApiKey
module.exports.signAndSend = signAndSend
